prow/plugins/dco/dco.go

```diff
@@ -0,0 +1,307 @@
+		return nil
+	}
+
+	if !hasNoLabel {
+		l.Infof("Adding %q label", dcoNoLabel)
+		// add 'dco-signoff: no' label
+		if err := gc.AddLabel(org, repo, pr.Number, dcoNoLabel); err != nil {
+			return fmt.Errorf("error adding label: %v", err)
+		}
+	}
+	if hasYesLabel {
+		l.Infof("Removing %q label", dcoYesLabel)
+		// remove 'dco-signoff: yes' label
+		if err := gc.RemoveLabel(org, repo, pr.Number, dcoYesLabel); err != nil {
+			return fmt.Errorf("error removing label: %v", err)
+		}
+	}
+	if existingStatus != github.StatusFailure {
+		l.Infof("Setting DCO status context to failed")
+		if err := gc.CreateStatus(org, repo, pr.Head.SHA, github.Status{
+			Context:     dcoContextName,
+			State:       github.StatusFailure,
+			TargetURL:   targetURL,
+			Description: dcoContextMessageFailed,
+		}); err != nil {
+			return fmt.Errorf("error setting pull request status: %v", err)
+		}
+	}
+
+	cp.PruneComments(shouldPrune(l, botName))
+	l.Infof("Commenting on PR to advise users of DCO check")
+	if err := gc.CreateComment(org, repo, pr.Number, fmt.Sprintf(dcoNotFoundMessage, plugins.AboutThisBot)); err != nil {
+		l.WithError(err).Warning("Could not create DCO not found comment.")
+	}
+
+	return nil
+}
+
+// 1. Check commit messages in the pull request for the sign-off string
+// 2. Check the existing status context value
+// 3. Check the existing PR labels
+// 4. If signed off, apply appropriate labels and status context.
+// 5. If not signed off, apply appropriate labels and status context and add a comment.
+func handle(gc gitHubClient, cp commentPruner, log *logrus.Entry, org, repo string, pr github.PullRequest) error {
+	l := log.WithField("pr", pr.Number)
+
+	signedOff, err := checkCommitMessages(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error running DCO check against commits in PR")
+		return err
+	}
+
+	existingStatus, err := checkExistingStatus(gc, l, org, repo, pr.Head.SHA)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR status")
+		return err
+	}
+
+	hasYesLabel, hasNoLabel, err := checkExistingLabels(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR labels")
+		return err
+	}
+
+	return takeAction(gc, cp, l, org, repo, pr, signedOff, existingStatus, hasYesLabel, hasNoLabel)
+}
+
+// shouldPrune finds comments left by this plugin.
+func shouldPrune(log *logrus.Entry, botName string) func(github.IssueComment) bool {
+	return func(comment github.IssueComment) bool {
+		if comment.User.Login != botName {
+			return false
+		}
+		return strings.Contains(comment.Body, dcoMsgPruneMatch)
+	}
+}
+
+func handlePullRequestEvent(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	org := pe.Repo.Owner.Login
+	repo := pe.Repo.Name
+
+	return handle(pc.GitHubClient, pc.CommentPruner, pc.Logger, org, repo, pe.PullRequest)
+}
+
+func handleCommentEvent(pc plugins.PluginClient, ce github.GenericCommentEvent) error {
```

Comment:
I have not written tests for `handleCommentEvent` nor `handlePullRequestEvent` as the bulk of the logic resides in `handle`.

I think this is probably okay for `handlePullRequestEvent`, but a rudimentary test for `handleCommentEvent` may be needed?